_posts/2018-02-12-Reusable-features-in-React-Redux.md

```diff
@@ -0,0 +1,188 @@
+Feel free to clone it and run it by **npm run start**. I will start by explaining how I approach creation
+of **standard react features** and later we'll get into topic of creating features that are able to create
+instances which can be mounted on different locations in the state. I call these features **reusable features**.
+
+## Standard features
+
+You can imagine a standard feature as a directory that lives in your codebase. 
+I will try to demonstrate what standard feature is on Todo List usecase. 
+The standard feature directory has prescribed structure and usually looks like this:
+
+```
+src/
+ todo-feature/
+   components/ 
+     TodoList.js     
+     TodoListItem.js
+   containers/
+     TodoList.js  
+     TodoListItem.js
+   action.js
+   reducers.js
+   selectors.js
+   utils.js
+   middleware.js
+```   
+
+#### components/
+
+The directory where your components live.
+
+#### containers/
+
+The directory where your containers live. If you don't know what is the difference between the components 
+and containers you should read [this article](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0)
+from Dan Abramov why this separation exists and why it is necessary.
+
+#### actions.js
+
+The file where your [action types](https://redux.js.org/docs/basics/Actions.html#actions) and
+[action creators](https://redux.js.org/docs/basics/Actions.html#action-creators) are defined.
+
+
+#### reducers.js
+
+The file where your [reducers](https://redux.js.org/docs/basics/Reducers.html) are defined. 
+This is also a place where you compose the reducers of your possible sub-features using [combineReducers](https://redux.js.org/docs/api/combineReducers.html).
+
+
+#### selectors.js
+
+The file where you define your selectors. Selectors are pure functions that compose and know how
+to retrieve data from your state. If you're now familiar with the concept I highly recommend looking 
+at [reselect](https://github.com/reactjs/reselect) library. I strongly recommend dividing the selectors into two categories. 
+Input-selectors that only retrieve data from the state and combined-selectors that can combine
+multiple input-selectors and add some business logic or do some transformations.
+
+
+#### utils.js
+
+Everything related to your feature that doesn't belong anywhere else. Usually it contains support utilities and helpers.
+
+#### middleware.js
+
+This is where our feature handles it's side effects, e.g. catching some Redux actions and dispatching new ones,
+synchronize the API calls, etc...
+
+You can check the [reference repository](https://github.com/char0n/react-reusable-features/tree/master/src/app/standard-feature) to see how this looks in a real project. 
+It depends on your use-case what the standard feature directory contains. 
+There are situations that you may need reducers but don't need actions and so on...
+But regardless, the full representation of the standard feature should look like that. 
```

Comment:
I'd remove the "but". We should avoid starting a sentence with this conjunction (https://www.dailywritingtips.com/can-you-start-sentences-with-“and”-and-“but”/).

---

> feature should look like that.

What is "that" ?